### PR TITLE
[Feature] Add DCFF

### PR DESCRIPTION
**Pull request by spynccat**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add DCFF method to pruning branch.

## Modification

1. Add FuseConv2d(dynamic ops).
2. Add DCFFChannelGroup(OneShotChannelGroup).
3. Add DCFFChannelMutator(BaseChannelMutator).
4. Add DCFF(BaseAlgorithm).
5. Add UT.
6. Add loss_calculator for mmdet, mmseg, mmpose.
7. Fix mmcv: dev-2.x -> mmcv: 2.x CI.
8. Fix pruning UT.
9. Typing modification for IndexDict unsupported in python3.6.9 typing.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: configs/pruning/mmcls/dcff/README.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+![pipeline](https://user-images.githubusercontent.com/31244134/189286581-722853ba-c6d7-4a39-b902-37995b444c71.jpg)
+
+## Results and models
```
Comment (sunnyxiaohu) [inline configs/pruning/mmcls/dcff/README.md:9]: The readME should add necessary notes on how to generate `resnet_cls.json`, `resnet_det.json`, and so on.

File: mmrazor/models/algorithms/pruning/dcff.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+VALID_MUTATOR_TYPE = Union[DCFFChannelMutator, Dict]
+VALID_DISTILLER_TYPE = Union[ConfigurableDistiller, Dict]
+VALID_PATH_TYPE = Union[str, Path]
+VALID_CHANNEL_CFG_PATH_TYPE = Union[VALID_PATH_TYPE, List[VALID_PATH_TYPE]]
+
+
+@MODELS.register_module()
+class DCFF(BaseAlgorithm):
+
+    def __init__(self,
```
Comment (sunnyxiaohu) [inline mmrazor/models/algorithms/pruning/dcff.py:27]: 1. add class docstring for DCFF
2. fuse count -> fuse_freq

File: mmrazor/models/algorithms/pruning/dcff.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            1 - torch.math.exp(-k * max_num))
+        T = A / (1 + torch.math.exp(-k * cur_num)) + t_s - A / 2
+        t = 1 / T
+        return t
+
+    def train_step(self, data: List[dict],
+                   optim_wrapper: OptimWrapper) -> Dict[str, torch.Tensor]:
+        """Train step."""
+
+        # self.message_hub = MessageHub.get_current_instance()
```
Comment (sunnyxiaohu) [inline mmrazor/models/algorithms/pruning/dcff.py:66]: clear

File: mmrazor/models/algorithms/pruning/dcff.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+from mmengine.model import BaseModel
+from mmengine.optim import OptimWrapper
+from torch import nn
+
+from mmrazor.models.distillers import ConfigurableDistiller
+from mmrazor.models.mutables import BaseMutable
+from mmrazor.models.mutators import DCFFChannelMutator
+from mmrazor.models.utils import reinitialize_optim_wrapper_count_status
+from mmrazor.registry import MODELS
+# from mmrazor.structures.subnet.fix_subnet import _dynamic_to_static
```
Comment (sunnyxiaohu) [inline mmrazor/models/algorithms/pruning/dcff.py:15]: clear

File: mmrazor/models/architectures/dynamic_ops/mixins/dynamic_conv_mixins.py [hunk 5]
```diff
@@ -404,3 +404,152 @@ def _get_dynamic_params_by_mutable_kernel_size(
...
+
+
+class FuseConvMixin(DynamicConvMixin):
+    """A mixin class for Pytorch conv, which can mutate ``in_channels``,
+    ``out_channels`` ."""
+
+    def forward_mixin(self: _ConvNd, x: Tensor) -> Tensor:
+        """Forward of dynamic conv2d OP."""
+        groups = self.groups
+        if self.groups == self.in_channels == self.out_channels:
```
Comment (sunnyxiaohu) [inline mmrazor/models/architectures/dynamic_ops/mixins/dynamic_conv_mixins.py:416]: Can L416-417 be deleted?

File: mmrazor/models/mutables/mutable_channel/units/dcff_channel_unit.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+
+@MODELS.register_module()
+class DCFFChannelUnit(OneShotMutableChannelUnit):
+    """``DCFFChannelUnit`` is for supernet DCFF and based on
+    OneShotMutableChannelUnit. In DCFF supernet, each module only has one
+    choice. The channel choice is fixed before training.
+
+    Args:
+        num_channels (int): The raw number of channels.
```
Comment (sunnyxiaohu) [inline mmrazor/models/mutables/mutable_channel/units/dcff_channel_unit.py:19]: update docstring

File: tests/test_models/test_algorithms/test_dcff_network.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    def _prepare_fake_data(self) -> Dict:
+        imgs = torch.randn(16, 3, 224, 224).to(self.device)
+        data_samples = [
+            ClsDataSample().set_gt_label(torch.randint(0, 1000,
+                                                       (16, ))).to(self.device)
+        ]
+
+        return {'inputs': imgs, 'data_samples': data_samples}
+
+    def prepare_dcff_model(self) -> DCFF:
```
Comment (sunnyxiaohu) [inline tests/test_models/test_algorithms/test_dcff_network.py:102]: What's the differences between `prepare_dcff_model` and `prepare_fixed_model`?

File: tests/test_models/test_mutables/test_units/test_dcff_channel_unit.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from unittest import TestCase
+
+from mmrazor.models.mutables import DCFFChannelUnit
+
+
+class TestDCFFChannelUnit(TestCase):
+
+    def test_init(self):
```
Comment (sunnyxiaohu) [inline tests/test_models/test_mutables/test_units/test_dcff_channel_unit.py:9]: Add more unit cases for DCFFChannelUnit.

File: mmrazor/models/mutators/channel_mutator/dcff_channel_mutator.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+    Args:
+        channel_cfgs (list[Dict]): A list of candidate channel configs.
+        channel_unit_cfg (Union[dict, Type[ChannelUnitType]], optional):
+            Config of MutableChannelUnits. Defaults to
+            dict( type='DCFFChannelUnit', units={}).
+    """
+
+    def __init__(self,
+                 channl_unit_cfg: Union[dict, Type[ChannelUnitType]] = dict(
```
Comment (sunnyxiaohu) [inline mmrazor/models/mutators/channel_mutator/dcff_channel_mutator.py:25]: 1. fix typo: replacing `channl` with `channel`
2. update docstring

File: mmrazor/models/mutables/mutable_channel/units/dcff_channel_unit.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        self._replace_with_dynamic_ops(
+            model, {
+                nn.Conv2d: dynamic_ops.FuseConv2d,
+                nn.BatchNorm2d: dynamic_ops.DynamicBatchNorm2d,
+                nn.Linear: dynamic_ops.DynamicLinear
+            })
+        self._register_channel_container(model, MutableChannelContainer)
+        self._register_mutable_channel(self.mutable_channel)
+
+    def alter_candidates_after_init(self, candidates: List[int]):
```
Comment (sunnyxiaohu) [inline mmrazor/models/mutables/mutable_channel/units/dcff_channel_unit.py:49]: Where will this method be used?

File: mmrazor/models/algorithms/pruning/dcff.py (hunk 2 above)
Comment (spynccat) [reply mmrazor/models/algorithms/pruning/dcff.py:47]: Updated.

File: mmrazor/models/mutables/mutable_channel/units/dcff_channel_unit.py (hunk 6 above)
Comment (spynccat) [reply mmrazor/models/mutables/mutable_channel/units/dcff_channel_unit.py:19]: Updated.

File: tests/test_models/test_mutables/test_units/test_dcff_channel_unit.py (hunk 8 above)
Comment (spynccat) [reply tests/test_models/test_mutables/test_units/test_dcff_channel_unit.py:17]: Updated.

File: mmrazor/models/mutators/channel_mutator/dcff_channel_mutator.py (hunk 9 above)
Comment (spynccat) [reply mmrazor/models/mutators/channel_mutator/dcff_channel_mutator.py]: Fixed.

File: mmrazor/models/architectures/dynamic_ops/mixins/dynamic_conv_mixins.py [hunk 11]
```diff
@@ -317,16 +317,6 @@ def _get_dynamic_params_by_mutable_kernel_size(
 
         return current_weight, current_padding
 
-    def forward_mixin(self: _ConvNd, x: Tensor) -> Tensor:
```
[the remark this answers is not in the document]
Comment (spynccat) [reply mmrazor/models/architectures/dynamic_ops/mixins/dynamic_conv_mixins.py:320]: Remove redundant forward_mixin() defined in base class DynamicConvMixin. In former commit get_dynamic_params() has different return var (modified).

File: mmrazor/models/mutators/channel_mutator/dcff_channel_mutator.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+        for key in unit_cfg:
+            assert num_subnets == len(
+                unit_cfg[key]['init_args']['candidate_choices'])
+            for i, value in enumerate(
+                    unit_cfg[key]['init_args']['candidate_choices']):
+                subnets[i][key] = value
+
+        return subnets
+
+    def calc_information(self, tau: float):
```
Comment (LKJacky) [inline mmrazor/models/mutators/channel_mutator/dcff_channel_mutator.py:78]: can move this code into dynamic op and expose an api for channelunit to set the number of channels.

Comment (LKJacky) [reply mmrazor/models/mutators/channel_mutator/dcff_channel_mutator.py:78]: dynamicop: forward and weight pruning
channelunit: operate mutablechannel
mutator: operate channelunit
algorithm: operate mutator

Comment (spynccat) [reply mmrazor/models/mutators/channel_mutator/dcff_channel_mutator.py:75]: Updated.

File: mmrazor/models/architectures/dynamic_ops/mixins/dynamic_conv_mixins.py [hunk 13]
```diff
@@ -404,3 +394,166 @@ def _get_dynamic_params_by_mutable_kernel_size(
...
+        _, topm_ids_order = torch.topk(
+            layeri_iscore_kl, int(real_out), sorted=False)
+        softmaxp = layeri_softmaxp[topm_ids_order, :]
+        if (not hasattr(self, 'layeri_softmaxp')):
+            setattr(self, 'layeri_softmaxp', softmaxp)
+        else:
+            self.layeri_softmaxp.data = softmaxp
+        del param, layeri_param, layeri_negaEudist, layeri_kl
+
+    def to_static_op(self: _ConvNd) -> nn.Conv2d:
```
[the remark this answers is not in the document]
Comment (spynccat) [reply mmrazor/models/architectures/dynamic_ops/mixins/dynamic_conv_mixins.py]: Fixed.

File: configs/pruning/mmcls/dcff/README.md [hunk 14]
```diff
@@ -0,0 +1,100 @@
+# Training Compact CNNs for Image Classification using Dynamic-coded Filter Fusion
```
[the remark this answers is not in the document]
Comment (spynccat) [reply configs/pruning/mmcls/dcff/README.md:1]: Updated. For convenience the empirical statistics are not split.

File: configs/pruning/mmcls/dcff/dcff_resnet_8xb32_in1k.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+            type='DCFFChannelUnit',
+            units='configs/pruning/mmcls/dcff/resnet_cls.json'),
+        parse_cfg=dict(
+            type='BackwardTracer',
+            loss_calculator=dict(type='ImageClassifierPseudoLoss'))))
+
+find_unused_parameters = True
+
+model_wrapper = dict(
+    type='mmcv.MMDistributedDataParallel', find_unused_parameters=True)
```
[the remark this answers is not in the document]
Comment (spynccat) [reply configs/pruning/mmcls/dcff/dcff_resnet_8xb32_in1k.py:43]: Fixed.

File: mmrazor/models/algorithms/pruning/dcff.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+@MODELS.register_module()
+class DCFF(BaseAlgorithm):
+    """DCFF Networks.
+
+    Please refer to paper
+    [Dynamic-coded Filter Fusion](https://arxiv.org/abs/2107.06916).
+
+    Args:
+        mutator (dict | :obj:`SlimmableChannelMutator`): The config of
```
[the remark this answers is not in the document]
Comment (spynccat) [reply mmrazor/models/algorithms/pruning/dcff.py]: Fixed.

File: mmrazor/models/algorithms/pruning/ite_prune_algorithm.py [hunk 17]
```diff
@@ -158,7 +163,7 @@ def group_target_pruning_ratio(
                     unit_target = target[unit_name]
                     assert isinstance(unit_target, (float, int))
                     group_target[group_id] = unit_target
-
+        print('group_target', group_target)
```
[the remark this answers is not in the document]
Comment (spynccat) [reply mmrazor/models/algorithms/pruning/ite_prune_algorithm.py]: Fixed.

File: mmrazor/models/algorithms/pruning/ite_prune_algorithm.py [hunk 18]
```diff
@@ -200,12 +209,37 @@ def _epoch(self):
 
     @property
     def _iteration(self):
-        """Get current iteration number."""
+        """Get total iteration number."""
```
[the remark this answers is not in the document]
Comment (spynccat) [reply mmrazor/models/algorithms/pruning/ite_prune_algorithm.py]: Updated docs, _iteration() for current sum iters, _current_iteration() for current iters in current epoch(if IterRunner set to 0)

File: mmrazor/models/mutators/channel_mutator/dcff_channel_mutator.py [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+                        },...
+                    }
+                Every unit must have the same number of candidate_choices, and
+                the candidate in the list of candidate_choices with the same
+                position compose a subnet.
+
+        Returns:
+            List[Dict[str, int]]: config of the subnets.
+        """
+        """Prepare subnet config."""
```
[the remark this answers is not in the document]
Comment (spynccat) [reply mmrazor/models/mutators/channel_mutator/dcff_channel_mutator.py]: Fixed.

File: mmrazor/models/mutables/mutable_channel/oneshot_mutable_channel.py [hunk 20]
```diff
@@ -35,6 +35,7 @@ def current_choice(self) -> Union[int, float]:
     @current_choice.setter
     def current_choice(self, choice: Union[int, float]):
         """Set current choice."""
+        print('choice', choice, self.candidate_choices)
         assert choice in self.candidate_choices
```
[the remark this answers is not in the document]
Comment (spynccat) [reply mmrazor/models/mutables/mutable_channel/oneshot_mutable_channel.py]: Fixed.

**Review by sunnyxiaohu — APPROVED** [review]
Next, handle depolyment pipeline in another PR

**Pull request merged**